File: frontend/services/gmail.ts
import { apiRequest } from "@/services/api-client";

export interface GmailConnection {
  oauth_configured: boolean;
  connected: boolean;
  gmail_address: string | null;
  scopes: string[];
  status: string;
  last_error: string | null;
  last_synced_at: string | null;
  token_expires_at: string | null;
  send_acknowledgment_emails: boolean;
}

interface GmailConnectionResponse {
  success: boolean;
  message: string;
  data: GmailConnection;
}

interface GmailAuthorizationResponse {
  success: boolean;
  message: string;
  data: {
    authorization_url: string;
  };
}

export interface GmailProcessResult {
  messages_scanned: number;
  messages_processed: number;
  attachments_stored: number;
  duplicates_skipped: number;
  unsupported_skipped: number;
  errors: number;
}

interface GmailProcessResponse {
  success: boolean;
  message: string;
  data: GmailProcessResult;
}

export async function getGmailConnection(): Promise<GmailConnection> {
  const response = await apiRequest<GmailConnectionResponse>("/gmail/status");
  return response.data;
}

export async function getGmailAuthorizationUrl(): Promise<string> {
  const response = await apiRequest<GmailAuthorizationResponse>(
    "/gmail/oauth/authorize",
  );
  return response.data.authorization_url;
}

export async function disconnectGmail(): Promise<void> {
  await apiRequest("/gmail/connection", { method: "DELETE" });
}

export async function updateGmailSettings(
  sendAcknowledgmentEmails: boolean,
): Promise<GmailConnection> {
  const response = await apiRequest<GmailConnectionResponse>("/gmail/settings", {
    method: "PATCH",
    body: JSON.stringify({
      send_acknowledgment_emails: sendAcknowledgmentEmails,
    }),
  });
  return response.data;
}

export async function processGmailInbox(): Promise<GmailProcessResult> {
  const response = await apiRequest<GmailProcessResponse>("/gmail/process", {
    method: "POST",
    body: JSON.stringify({ max_messages: 25 }),
  });
  return response.data;
}

File: backend/app/schemas/gmail.py
from datetime import datetime

from pydantic import BaseModel, Field, HttpUrl


class GmailConnection(BaseModel):
    oauth_configured: bool
    connected: bool
    gmail_address: str | None = None
    scopes: list[str] = Field(default_factory=list)
    status: str = "disconnected"
    last_error: str | None = None
    last_synced_at: datetime | None = None
    token_expires_at: datetime | None = None
    send_acknowledgment_emails: bool = False


class GmailConnectionResponse(BaseModel):
    success: bool = True
    message: str
    data: GmailConnection


class GmailAuthorization(BaseModel):
    authorization_url: HttpUrl


class GmailAuthorizationResponse(BaseModel):
    success: bool = True
    message: str
    data: GmailAuthorization


class GmailDisconnectResponse(BaseModel):
    success: bool = True
    message: str


class GmailSettingsUpdate(BaseModel):
    send_acknowledgment_emails: bool


class GmailProcessRequest(BaseModel):
    max_messages: int = Field(default=25, ge=1, le=100)


class GmailProcessResult(BaseModel):
    messages_scanned: int = 0
    messages_processed: int = 0
    attachments_stored: int = 0
    duplicates_skipped: int = 0
    unsupported_skipped: int = 0
    errors: int = 0


class GmailProcessResponse(BaseModel):
    success: bool = True
    message: str
    data: GmailProcessResult
